src/interactivity/base/HostSignalInputThread.cpp

```diff
@@ -0,0 +1,240 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "precomp.h"
+
+#include "HostSignalInputThread.hpp"
+#include "../inc/HostSignals.hpp"
+
+#include "../interactivity/inc/ServiceLocator.hpp"
+
+using namespace Microsoft::Console;
+using namespace Microsoft::Console::Interactivity;
+
+// Constructor Description:
+// - Creates the PTY Signal Input Thread.
+// Arguments:
+// - hPipe - a handle to the file representing the read end of the Host Signal pipe.
+HostSignalInputThread::HostSignalInputThread(_In_ wil::unique_hfile&& hPipe) :
+    _hFile{ std::move(hPipe) },
+    _hThread{},
+    _dwThreadId{ 0 },
+    _consoleConnected{ false }
+{
+    THROW_HR_IF(E_HANDLE, _hFile.get() == INVALID_HANDLE_VALUE);
+}
+
+HostSignalInputThread::~HostSignalInputThread()
+{
+    // Manually terminate our thread during unittesting. Otherwise, the test
+    //      will finish, but TAEF will not manually kill the test.
+#ifdef UNIT_TESTING
+    TerminateThread(_hThread.get(), 0);
+#endif
+}
+
+// Function Description:
+// - Static function used for initializing an instance's ThreadProc.
+// Arguments:
+// - lpParameter - A pointer to the HostSignalInputThread instance that should be called.
+// Return Value:
+// - The return value of the underlying instance's _InputThread
+DWORD WINAPI HostSignalInputThread::StaticThreadProc(_In_ LPVOID lpParameter)
+{
+    HostSignalInputThread* const pInstance = reinterpret_cast<HostSignalInputThread*>(lpParameter);
+    return pInstance->_InputThread();
+}
+
+// Method Description:
+// - Tell us that there's a client attached to the console, so we can actually
+//      do something with the messages we receive now. Before this is set, there
+//      is no guarantee that a client has attached, so most parts of the console
+//      (in and screen buffers) haven't yet been initialized.
+// Arguments:
+// - <none>
+// Return Value:
+// - <none>
+void HostSignalInputThread::ConnectConsole() noexcept
+{
+    _consoleConnected = true;
+}
+
+// Method Description:
+// - Attempts to retrieve a given type T off of the internal
+//   pipe channel and return it.
+// Return Value:
+// - A structure filled with the specified data off the byte stream
+// - EXCEPTIONS may be thrown if the packet size mismatches
+//     or if we fail to read for another reason.
+template<typename T>
+T HostSignalInputThread::_ReceiveTypedPacket()
+{
+    T msg = { 0 };
+    THROW_HR_IF(E_ABORT, !_GetData(&msg, sizeof(msg)));
+
+    // If the message size was stated to be larger, we
+    // want to seek forward to the next message code.
+    // If it's equal, we'll seek forward by 0 and
+    // do nothing.
+    if (msg.sizeInBytes >= sizeof(msg))
+    {
+        _AdvanceReader(msg.sizeInBytes - sizeof(msg));
+    }
+    // If the message is smaller than what we expected
+    // then it was malformed and we need to throw.
+    else
+    {
+        THROW_HR(E_ILLEGAL_METHOD_CALL);

```

Comment:
If you swap the if condition around you can unnest the `_AdvanceReader` and make the code flow more linear / less nested.